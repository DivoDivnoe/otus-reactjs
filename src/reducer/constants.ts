export enum ActionType {
  SET_USER = 'SET_USER',
  RESET_USER = 'RESET_USER',
  START_PLAYING = 'START_PLAYING',
  STOP_PLAYING = 'STOP_PLAYING',
  SET_SIZE = 'SET_SIZE',
  SET_SPEED = 'SET_SPEED',
  SET_FILL = 'SET_FILL',
  SET_MODEL = 'SET_MODEL',
  RESET_MODEL = 'RESET_MODEL',
}
