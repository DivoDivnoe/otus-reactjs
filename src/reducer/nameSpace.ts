enum NameSpace {
  GAME = 'game',
  USER = 'user',
}

export default NameSpace;
