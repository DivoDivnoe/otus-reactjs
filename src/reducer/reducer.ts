import { combineReducers } from 'redux';

import gameReducer, {
  GameState,
  NAME_SPACE as GAME_NAME_SPACE,
} from '@/modules/game';
import userReducer, {
  UserState,
  NAME_SPACE as USER_NAME_SPACE,
} from '@/modules/user';

export interface State {
  [GAME_NAME_SPACE]: GameState;
  [USER_NAME_SPACE]: UserState;
}

const reducer = combineReducers({
  [GAME_NAME_SPACE]: gameReducer,
  [USER_NAME_SPACE]: userReducer,
});

export default reducer;
