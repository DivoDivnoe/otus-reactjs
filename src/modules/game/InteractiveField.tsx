import React, { FC, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Field } from '@/components/Field';
import { State } from '@/reducer';
import { BoardSize, getSize } from './size';
import {
  Model,
  getModel,
  ActionCreator as ModelActionCreator,
  Coords,
} from './model';

export const InteractiveField: FC = () => {
  const dispatch = useDispatch();

  const size = useSelector<State, BoardSize>(getSize);
  const model = useSelector<State, Model>(getModel);

  const onClickCell = useCallback(
    (coords: Coords): void => {
      dispatch(ModelActionCreator.updateModel(coords));
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  return <Field size={size} model={model} clickHandler={onClickCell} />;
};
