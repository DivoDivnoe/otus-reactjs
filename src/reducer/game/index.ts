import { combineReducers } from 'redux';
import ModelReducer from './model';
import IsPlayingReducer from './isPlaying';
import SizeReducer from './size';
import SpeedReducer from './speed';
import FillReducer from './fill';
import { Model } from '@/core';
import { BoardSize, SpeedType, FillType } from '@/constants';

export interface GameState {
  size: BoardSize;
  speed: SpeedType;
  fill: FillType;
  isPlaying: boolean;
  model: Model;
}

const reducer = combineReducers({
  size: SizeReducer,
  speed: SpeedReducer,
  fill: FillReducer,
  isPlaying: IsPlayingReducer,
  model: ModelReducer,
});

export default reducer;
