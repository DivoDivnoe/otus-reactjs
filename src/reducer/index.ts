import reducer from './reducer';

export { State } from './reducer';
export default reducer;
