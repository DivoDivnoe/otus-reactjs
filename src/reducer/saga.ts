import {
  call,
  put,
  takeEvery,
  fork,
  select,
  StrictEffect,
} from 'redux-saga/effects';
import { APP_KEY } from './constants';
import { State } from '@/reducer';
import { ActionCreator as UserActionCreator, getUser } from '@/modules/user';
import {
  ActionCreator as ModelActionCreator,
  getModel,
  Model,
} from '@/modules/game/model';
import {
  ActionCreator as SizeActionCreator,
  getSize,
  BoardSize,
} from '@/modules/game/size';
import {
  ActionCreator as SpeedActionCreator,
  getSpeed,
} from '@/modules/game/speed';

import {
  ActionCreator as FillActionCreator,
  getFill,
  FillType,
} from '@/modules/game/fill';

import {
  ActionCreator as IsPlayingActionCreator,
  getIsPlaying,
} from '@/modules/game/isPlaying';
import { createRandomMatrix } from '@/modules/game/core';

export const getFromLocalStorage = (key: string): string | null => {
  return window.localStorage.getItem(key);
};

export const saveToLocalStorage = (key: string, state: string): void => {
  window.localStorage.setItem(key, state);
};

export function* createModel(): Generator<
  StrictEffect,
  void,
  Model | State | BoardSize | FillType
> {
  const size = yield select(getSize);
  const fill = yield select(getFill);

  const model = yield call(
    createRandomMatrix,
    size as BoardSize,
    fill as FillType
  );
  yield put(ModelActionCreator.setModel(model as Model));
}

export function* getStateFromLocalStorage(): Generator<
  StrictEffect,
  void,
  string | null | State
> {
  const rawState = yield call(getFromLocalStorage, APP_KEY);
  let state: State;

  if (rawState) {
    state = JSON.parse(rawState as string);

    yield put(UserActionCreator.signin(getUser(state)));
    yield put(SizeActionCreator.setSize(getSize(state)));
    yield put(SpeedActionCreator.setSpeed(getSpeed(state)));
    yield put(FillActionCreator.setFill(getFill(state)));
    yield put(IsPlayingActionCreator.setPlaying(getIsPlaying(state)));
    yield put(ModelActionCreator.setModel(getModel(state)));
  } else {
    yield fork(createModel);
  }
}

export function* saveStateToLocalStorage(): Generator<
  StrictEffect,
  void,
  State
> {
  const state: State = yield select();
  const serializedState = JSON.stringify(state);

  yield call(saveToLocalStorage, APP_KEY, serializedState);
}

export function* actionsWatcher(): Generator<StrictEffect, void, void> {
  const actions = [
    UserActionCreator.signin.type,
    UserActionCreator.signout.type,
    SizeActionCreator.setSize.type,
    SpeedActionCreator.setSpeed.type,
    FillActionCreator.setFill.type,
    ModelActionCreator.setModel.type,
    ModelActionCreator.resetModel.type,
    IsPlayingActionCreator.setPlaying.type,
    IsPlayingActionCreator.startPlaying.type,
    IsPlayingActionCreator.stopPlaying.type,
  ];

  for (const action of actions) {
    yield takeEvery(action, saveStateToLocalStorage);
  }

  yield takeEvery(SizeActionCreator.setSize.type, createModel);
  yield takeEvery(FillActionCreator.setFill.type, createModel);
}

export function* stateSaga(): Generator<StrictEffect, void, void> {
  yield fork(getStateFromLocalStorage);
  yield fork(actionsWatcher);
}
