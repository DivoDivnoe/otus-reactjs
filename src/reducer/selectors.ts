import { UserState, NAME_SPACE as USER_NAME_SPACE } from '@/modules/user';
import { GameState, NAME_SPACE as GAME_NAME_SPACE } from '@/modules/game';
import { State } from '@/reducer';

export const getUserState = (state: State): UserState => state[USER_NAME_SPACE];
export const getGameState = (state: State): GameState => state[GAME_NAME_SPACE];
