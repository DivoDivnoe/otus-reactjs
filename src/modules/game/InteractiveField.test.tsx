import React from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useSelector, useDispatch } from 'react-redux';
import { InteractiveField } from './InteractiveField';
import reducer from '@/reducer';
import { act } from 'react-dom/test-utils';

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('InteractiveField test cases', () => {
  it('InteractiveField check', () => {
    const mockDispatch = jest.fn();

    (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
    (useSelector as jest.Mock).mockImplementation((callback) => {
      return callback(store.getState());
    });

    const store = configureStore({ reducer });
    act(() => {
      store.dispatch({
        type: 'model/setModel',
        payload: [
          [0, 0],
          [0, 0],
        ],
      });
    });

    const { asFragment } = render(
      <Provider store={store}>
        <InteractiveField />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();

    userEvent.click(screen.getAllByTestId('cell')[0]);
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'model/updateModel',
      payload: { x: 0, y: 0 },
    });
  });
});
