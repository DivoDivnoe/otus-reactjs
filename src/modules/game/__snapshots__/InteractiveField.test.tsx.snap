// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractiveField test cases InteractiveField check 1`] = `
<DocumentFragment>
  <div
    class="css-avf8ql"
  >
    <div
      class="css-9ncwx"
    >
      <div
        class="css-1dept0h"
        data-testid="cell"
      />
      <div
        class="css-1dept0h"
        data-testid="cell"
      />
    </div>
    <div
      class="css-9ncwx"
    >
      <div
        class="css-1dept0h"
        data-testid="cell"
      />
      <div
        class="css-1dept0h"
        data-testid="cell"
      />
    </div>
  </div>
</DocumentFragment>
`;
